File: DSA/Ishita-Nigam/Day14/Day14_Q3.java
import java.util.Scanner;

public class Day14_Q3 {

    public static int minBulbsToLightHallway(int n, int[] hallway) {
        int[] maxRange = new int[n];

        for (int i = 0; i < n; i++) {
            if (hallway[i] == -1) {
                maxRange[i] = -1;
            } else {
                int left = Math.max(0, i - hallway[i]);
                int right = Math.min(n - 1, i + hallway[i]);
                for (int j = left; j <= right; j++) {
                    maxRange[j] = Math.max(maxRange[j], right);
                }
            }
        }

        int bulbsOn = 0;
        int currentEnd = 0;
        int farthest = 0;

        for (int i = 0; i < n; i++) {
            if (maxRange[i] == -1) {
                return -1;
            }
            farthest = Math.max(farthest, maxRange[i]);
            if (i == currentEnd) {
                bulbsOn++;
                currentEnd = farthest;
                if (currentEnd >= n - 1) {
                    break;
                }
            }
        }

        return currentEnd >= n - 1 ? bulbsOn : -1;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter the number of partitions: ");
        int n = scanner.nextInt();

        int[] hallway = new int[n];
        System.out.println("Enter the power of bulbs at each partition:");
        for (int i = 0; i < n; i++) {
            hallway[i] = scanner.nextInt();
        }

        int result = minBulbsToLightHallway(n, hallway);
        System.out.println("Minimum number of bulbs to light up the hallway: " + result);

        scanner.close();
    }
}

File: DSA/Hima-Swarupa-Padala/day14q3.py
def min_bulbs_to_light_hallway(n, hallway):
   
    coverage = []
    
    for i in range(n):
        if hallway[i] != -1:
            start = max(0, i - hallway[i])
            end = min(n - 1, i + hallway[i])
            coverage.append((start, end))

    coverage.sort(key=lambda x: (x[0], x[1]))
    
    num_bulbs = 0
    current_end = 0
    i = 0
    
    while i < len(coverage) and current_end < n:
        
        if coverage[i][0] > current_end:
            return -1  
        
        max_end = current_end
        while i < len(coverage) and coverage[i][0] <= current_end:
            max_end = max(max_end, coverage[i][1])
            i += 1

        num_bulbs += 1
        current_end = max_end + 1  

    if current_end >= n:
        return num_bulbs
    else:
        return -1

n = 6
hallway = [-1, 2, 2, -1, 0, 0]
print(min_bulbs_to_light_hallway(n, hallway))  

def min_bulbs_to_turn_on(n, hallway):
    bulbs_on = 0
    last_bulb = -1
    i = 0
    
    while i < n:
        if hallway[i] == -1:
            i += 1
            continue
        
        start = max(i - hallway[i], 0)
        end = min(i + hallway[i], n - 1)
        
        if start > last_bulb:
            return -1
        
        bulbs_on += 1
        last_bulb = end
        i = end + 1
    
    return bulbs_on

n = 9
hallway2 = [2, 3, 4, -1, 2, 0, 0, -1, 0]
print(min_bulbs_to_turn_on(n, hallway2))  
